File: game-core/src/main/kotlin/com/thevaguebox/picpac/core/ai/SearchModel.kt
package com.thevaguebox.picpac.core.ai

import com.thevaguebox.picpac.core.Board
import com.thevaguebox.picpac.core.Cell
import com.thevaguebox.picpac.core.Symbol

data class PicPacDecisionState(
    val board: Board,
    val heldSymbol: Symbol,
    val remainingX: Int,
    val remainingO: Int,
) {
    init {
        require(remainingX in 0..5 && remainingO in 0..5)
        require(board.count(Symbol.X) + remainingX + (if (heldSymbol == Symbol.X) 1 else 0) == 5)
        require(board.count(Symbol.O) + remainingO + (if (heldSymbol == Symbol.O) 1 else 0) == 5)
        require(!board.hasWinner() && !board.isFull)
    }

    val legalCells: List<Cell> get() = board.legalCells()
    val hiddenTotal: Int get() = remainingX + remainingO
}

data class PicPacChanceState(
    val board: Board,
    val remainingX: Int,
    val remainingO: Int,
) {
    val total: Int get() = remainingX + remainingO
}

sealed interface SearchTransition {
    data object Win : SearchTransition
    data object Draw : SearchTransition
    data class Chance(val state: PicPacChanceState) : SearchTransition
}

object PublicPicPacSearchModel {
    fun from(observation: AiObservation): PicPacDecisionState = PicPacDecisionState(
        board = observation.board,
        heldSymbol = observation.heldSymbol,
        remainingX = observation.remainingX,
        remainingO = observation.remainingO,
    )

    fun place(state: PicPacDecisionState, cell: Cell): SearchTransition {
        require(state.board[cell] == null)
        val board = state.board.place(cell, state.heldSymbol)
        if (board.winningLines(state.heldSymbol).isNotEmpty()) return SearchTransition.Win
        if (board.isFull) return SearchTransition.Draw
        return SearchTransition.Chance(
            PicPacChanceState(board, state.remainingX, state.remainingO),
        )
    }

    fun draw(state: PicPacChanceState, symbol: Symbol): PicPacDecisionState {
        require(state.total > 0)
        require(if (symbol == Symbol.X) state.remainingX > 0 else state.remainingO > 0)
        return PicPacDecisionState(
            board = state.board,
            heldSymbol = symbol,
            remainingX = state.remainingX - if (symbol == Symbol.X) 1 else 0,
            remainingO = state.remainingO - if (symbol == Symbol.O) 1 else 0,
        )
    }
}

File: ios/Packages/PicPacKit/Sources/PicPacAI/SearchModel.swift
import PicPacCore

public enum SearchModelError: Error, Equatable, Sendable {
    case invalidState
    case occupiedCell(Int)
    case emptyBag
    case exhaustedSymbol(Symbol)
}

public struct PicPacDecisionState: Equatable, Hashable, Sendable {
    public let board: Board
    public let heldSymbol: Symbol
    public let remainingX: Int
    public let remainingO: Int

    public init(
        board: Board,
        heldSymbol: Symbol,
        remainingX: Int,
        remainingO: Int
    ) throws {
        guard (0...5).contains(remainingX), (0...5).contains(remainingO),
              board.count(of: .x) + remainingX + (heldSymbol == .x ? 1 : 0) == 5,
              board.count(of: .o) + remainingO + (heldSymbol == .o ? 1 : 0) == 5,
              !board.hasWinner,
              !board.isFull else {
            throw SearchModelError.invalidState
        }
        self.board = board
        self.heldSymbol = heldSymbol
        self.remainingX = remainingX
        self.remainingO = remainingO
    }

    public var legalCells: [Cell] { board.legalCells }
    public var hiddenTotal: Int { remainingX + remainingO }
}

public struct PicPacChanceState: Equatable, Hashable, Sendable {
    public let board: Board
    public let remainingX: Int
    public let remainingO: Int

    public init(board: Board, remainingX: Int, remainingO: Int) throws {
        guard (0...5).contains(remainingX), (0...5).contains(remainingO),
              board.count(of: .x) + remainingX == 5,
              board.count(of: .o) + remainingO == 5,
              remainingX + remainingO > 0,
              !board.hasWinner,
              !board.isFull else {
            throw SearchModelError.invalidState
        }
        self.board = board
        self.remainingX = remainingX
        self.remainingO = remainingO
    }

    public var total: Int { remainingX + remainingO }
}

public enum SearchTransition: Equatable, Hashable, Sendable {
    case win
    case draw
    case chance(PicPacChanceState)
}

public enum PublicPicPacSearchModel {
    public static func from(_ observation: AiObservation) -> PicPacDecisionState {
        try! PicPacDecisionState(
            board: observation.board,
            heldSymbol: observation.heldSymbol,
            remainingX: observation.remainingX,
            remainingO: observation.remainingO
        )
    }

    public static func place(
        _ state: PicPacDecisionState,
        at cell: Cell
    ) throws -> SearchTransition {
        guard state.board[cell] == nil else {
            throw SearchModelError.occupiedCell(cell.index)
        }
        let board = try state.board.placing(state.heldSymbol, at: cell)
        if !board.winningLines(for: state.heldSymbol).isEmpty { return .win }
        if board.isFull { return .draw }
        return .chance(
            try PicPacChanceState(
                board: board,
                remainingX: state.remainingX,
                remainingO: state.remainingO
            )
        )
    }

    public static func draw(
        _ state: PicPacChanceState,
        symbol: Symbol
    ) throws -> PicPacDecisionState {
        guard state.total > 0 else { throw SearchModelError.emptyBag }
        let available = symbol == .x ? state.remainingX : state.remainingO
        guard available > 0 else { throw SearchModelError.exhaustedSymbol(symbol) }
        return try PicPacDecisionState(
            board: state.board,
            heldSymbol: symbol,
            remainingX: state.remainingX - (symbol == .x ? 1 : 0),
            remainingO: state.remainingO - (symbol == .o ? 1 : 0)
        )
    }
}
